### PR TITLE
JENKINS-59320 - Should log all running jobs when bundle is generated

### PR DESCRIPTION
**Pull request by Evildethow**
https://issues.jenkins-ci.org/browse/JENKINS-59320

Replaces https://github.com/jenkinsci/support-core-plugin/pull/188

### CODE REVIEW

**Comment by varyvol** [comment]
I think the anonymization concerns that were raised in the previous PR still apply.

**Comment by aheritier** [comment]
yes good point @varyvol URLs should be anomymized @Evildethow if the feature is active

File: src/main/java/com/cloudbees/jenkins/support/impl/RunningJobs.java [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+                @Override
+                public void writeTo(OutputStream os) throws IOException {
+                    try (OutputStreamWriter osw = new OutputStreamWriter(os, StandardCharsets.UTF_8); BufferedWriter bw = new BufferedWriter(osw); PrintWriter out = new PrintWriter(bw)) {
+                        Optional.ofNullable(Jenkins.getInstanceOrNull())
+                            .ifPresent(jenkins -> Arrays.stream(jenkins.getComputers())
+                                .flatMap(computer -> computer.getExecutors().stream())
+                                .filter(Executor::isBusy)
+                                .collect(Collectors.toList())
+                                .forEach(executor -> Optional.ofNullable(executor.getCurrentExecutable())
+                                    .ifPresent(executable -> out.println(executable.getParent().getOwnerTask().getUrl()))));
```
Comment (MRamonLeon) [inline src/main/java/com/cloudbees/jenkins/support/impl/RunningJobs.java]: Well, it really got filtered without doing anything. It's not a prefiltered content, but it doesn't matter. Maybe for performance reasons, it's good to be a pre-filtered content, but it works, IIRC.

File: src/main/java/com/cloudbees/jenkins/support/impl/RunningJobs.java [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+    @Override
+    @NonNull
+    public String getDisplayName() {
+        return "Running Jobs";
+    }
+
+    @Override
+    public void addContents(@NonNull Container result) {
+        result.add(
+            new Content("nodes/master/running-jobs.txt") {
```
Comment (MRamonLeon) [inline src/main/java/com/cloudbees/jenkins/support/impl/RunningJobs.java]: A **PrintedContent** makes it easier.

File: src/test/java/com/cloudbees/jenkins/support/impl/RunningJobsTest.java [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+        String jobName = "running-job";
+        FreeStyleProject p = j.createFreeStyleProject(jobName);
+        QueueTaskFuture<FreeStyleBuild> build = p.scheduleBuild2(0);
+        build.waitForStart();
+
+        String output = SupportTestUtils.invokeComponentToString(new RunningJobs());
+
+        assertThat(output, containsString(p.getUrl()));
+    }
+}
```
Comment (MRamonLeon) [inline src/test/java/com/cloudbees/jenkins/support/impl/RunningJobsTest.java]: nit: lack the new line

File: src/main/java/com/cloudbees/jenkins/support/impl/RunningJobs.java [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+                @Override
+                public void writeTo(OutputStream os) throws IOException {
+                    try (OutputStreamWriter osw = new OutputStreamWriter(os, StandardCharsets.UTF_8); BufferedWriter bw = new BufferedWriter(osw); PrintWriter out = new PrintWriter(bw)) {
+                        Optional.ofNullable(Jenkins.getInstanceOrNull())
+                            .ifPresent(jenkins -> Arrays.stream(jenkins.getComputers())
+                                .flatMap(computer -> computer.getExecutors().stream())
+                                .filter(Executor::isBusy)
```
Comment (jglick) [inline src/main/java/com/cloudbees/jenkins/support/impl/RunningJobs.java]: This method is close to useless. Several years ago, @kohsuke rewrote `Executor` to create a fresh one for each task, so basically every executor is busy almost as soon as it is created.

File: src/main/java/com/cloudbees/jenkins/support/impl/RunningJobs.java [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+    @Override
+    public void addContents(@NonNull Container result) {
+        result.add(
+            new Content("nodes/master/running-jobs.txt") {
+                @Override
+                public void writeTo(OutputStream os) throws IOException {
+                    try (OutputStreamWriter osw = new OutputStreamWriter(os, StandardCharsets.UTF_8); BufferedWriter bw = new BufferedWriter(osw); PrintWriter out = new PrintWriter(bw)) {
+                        Optional.ofNullable(Jenkins.getInstanceOrNull())
+                            .ifPresent(jenkins -> Arrays.stream(jenkins.getComputers())
+                                .flatMap(computer -> computer.getExecutors().stream())
```
Comment (jglick) [inline src/main/java/com/cloudbees/jenkins/support/impl/RunningJobs.java]: As previously mentioned, this will _not_ include Pipeline builds. You need `getAllExecutors`.

(Then you _may_ want to exclude, or specially display in a nested format, `node` blocks which are owned by the build overall. Unfortunately there is no equivalent of `SubTask.getOwnerTask` for `Executable`, only `getParent`, so from a `PlaceholderTask` you can only infer the owning build via `Task.url` being equal to an `Executable.url`.)

File: src/main/java/com/cloudbees/jenkins/support/impl/RunningJobs.java (hunk 1 above)
Comment (jglick) [inline src/main/java/com/cloudbees/jenkins/support/impl/RunningJobs.java]: This does not make much sense. You are printing the job name when you could be printing the specific build. And there is no deduplication, and the format is URL-oriented which is less than readable. So if three builds are running for one branch of a multibranch project, you will see something like

```
job/some-repo/job/master/
job/some-repo/job/master/
job/some-repo/job/master/
```

rather than what you might expect

```
some-repo/master#112
some-repo/master#113
some-repo/master#114
```

Suggest, as I did before:

```suggestion
                                    .ifPresent(executable -> out.println(executable))));
```

File: src/test/java/com/cloudbees/jenkins/support/impl/RunningJobsTest.java [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+
+public class RunningJobsTest {
+
+    @Rule
+    public JenkinsRule j = new JenkinsRule();
+
+    @Test
+    public void addContents() throws Exception {
+        String jobName = "running-job";
+        FreeStyleProject p = j.createFreeStyleProject(jobName);
```
Comment (jglick) [inline src/test/java/com/cloudbees/jenkins/support/impl/RunningJobsTest.java]: Really need to import the BOM, add Pipeline test dependencies, and verify behavior for a Pipeline build running in a couple of `node` blocks in `parallel` to see what this would look like.

File: src/main/java/com/cloudbees/jenkins/support/impl/RunningJobs.java (hunk 4 above)
Comment (jglick) [reply src/main/java/com/cloudbees/jenkins/support/impl/RunningJobs.java]: https://github.com/jenkinsci/jenkins/pull/944 FTR

File: src/main/java/com/cloudbees/jenkins/support/impl/RunningJobs.java (hunk 1 above)
Comment (jglick) [reply src/main/java/com/cloudbees/jenkins/support/impl/RunningJobs.java]: …though it looks like this will not currently print friendly content for `PlaceholderExecutable`s. Based on my reading of `Executable.toString` Javadoc, this would be a bug in `workflow-durable-task-step-plugin`, easy enough to fix.

Comment (Evildethow) [reply src/main/java/com/cloudbees/jenkins/support/impl/RunningJobs.java]: @MRamonLeon its not clear to me what I have to do to add anonymization support. Can you provide what steps are required? Also, it might be useful to document how to use this feature.

File: src/main/java/com/cloudbees/jenkins/support/impl/RunningJobs.java (hunk 4 above)
Comment (Evildethow) [reply src/main/java/com/cloudbees/jenkins/support/impl/RunningJobs.java]: @jglick I don't understand what you are suggesting that I do here. How do I filter executors if this method cannot be trusted?

File: src/main/java/com/cloudbees/jenkins/support/impl/RunningJobs.java (hunk 1 above)
Comment (Evildethow) [reply src/main/java/com/cloudbees/jenkins/support/impl/RunningJobs.java]: @jglick I don't understand what you are suggesting I do here.

File: src/test/java/com/cloudbees/jenkins/support/impl/RunningJobsTest.java (hunk 6 above)
Comment (Evildethow) [reply src/test/java/com/cloudbees/jenkins/support/impl/RunningJobsTest.java]: @jglick I assume you are referring to https://github.com/jenkinsci/bom

Do you have an example of what you are asking me to do? I'd rather not play PR ping pong trying to guess.

File: src/main/java/com/cloudbees/jenkins/support/impl/RunningJobs.java (hunk 1 above)
Comment (Evildethow) [reply src/main/java/com/cloudbees/jenkins/support/impl/RunningJobs.java]: Resolved by filtering PlaceholderExecutables

```
.forEach(executor -> Optional.ofNullable(executor.getCurrentExecutable())
                                .filter(executable -> !(executable.getParent() instanceof ExecutorStepExecution.PlaceholderTask))
```

File: src/test/java/com/cloudbees/jenkins/support/impl/RunningJobsTest.java (hunk 6 above)
Comment (Evildethow) [reply src/test/java/com/cloudbees/jenkins/support/impl/RunningJobsTest.java]: imported BOM and added a Pipeline test.

NOTE: Required bumping required from 2.107 -> 2.138.4

File: src/main/java/com/cloudbees/jenkins/support/impl/RunningJobs.java (hunk 2 above)
Comment (Evildethow) [reply src/main/java/com/cloudbees/jenkins/support/impl/RunningJobs.java]: Using PrefilteredPrintedContent

File: src/main/java/com/cloudbees/jenkins/support/impl/RunningJobs.java (hunk 1 above)
Comment (Evildethow) [reply src/main/java/com/cloudbees/jenkins/support/impl/RunningJobs.java]: Using PrefilteredPrintedContent

**Comment by Evildethow** [comment]
@jglick @MRamonLeon thanks for the review. I've made the changes to the best of my ability.

@aheritier NOTE: To get this working w/ Pipeline the base version had to move from 2.107 -> 2.138.4, and we also need to pull in workflow-durable-task-step as an optional dep.

Thanks, all

File: src/main/java/com/cloudbees/jenkins/support/impl/AboutJenkins.java [hunk 7]
```diff
@@ -510,7 +510,7 @@ private static String humanReadableSize(long size) {
             }
             final JenkinsLocationConfiguration jlc = JenkinsLocationConfiguration.get();
 
-            out.println("  * Url:     " + (jlc != null ? ContentFilter.filter(filter, jlc.getUrl()) : "No JenkinsLocationConfiguration available"));
+            out.println("  * Url:     " + (ContentFilter.filter(filter, jlc.getUrl())));
```
Comment (aheritier) [inline src/main/java/com/cloudbees/jenkins/support/impl/AboutJenkins.java]: Why did you remove this Null test @Evildethow  ? (I don't know that code thus maybe it's useless..)

Comment (jglick) [reply src/main/java/com/cloudbees/jenkins/support/impl/AboutJenkins.java]: `get` is `@Nonnull`.

File: src/test/java/com/cloudbees/jenkins/support/impl/RunningJobsTest.java (hunk 6 above)
Comment (jglick) [reply src/test/java/com/cloudbees/jenkins/support/impl/RunningJobsTest.java]: To be clear, the BOM part is just a convenience, but anyway there is little reason to still be putting out releases for very old LTS lines.

File: pom.xml [hunk 8]
```diff
@@ -88,6 +88,18 @@
     </pluginRepository>
   </pluginRepositories>
 
+  <dependencyManagement>
+    <dependencies>
+      <dependency>
+        <groupId>io.jenkins.tools.bom</groupId>
+        <artifactId>bom</artifactId>
+        <version>2.138.1</version>
```
Comment (jglick) [inline pom.xml]:
```suggestion
        <version>2.138.2</version>
```

File: pom.xml [hunk 9]
```diff
@@ -141,6 +153,22 @@
       <version>0.25</version>
       <scope>test</scope>
     </dependency>
+    <!-- Managed by BOM -->
```
Comment (jglick) [inline pom.xml]: Note that you _could_ also remove version overrides for some other plugins, such as `credentials`.

File: src/main/java/com/cloudbees/jenkins/support/impl/RunningJobs.java [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+    @NonNull
+    @Override
+    public Set<Permission> getRequiredPermissions() {
+        return Collections.singleton(Jenkins.ADMINISTER);
+    }
+
+    @Override
+    @NonNull
+    public String getDisplayName() {
+        return "Running Jobs";
```
Comment (jglick) [inline src/main/java/com/cloudbees/jenkins/support/impl/RunningJobs.java]:
```suggestion
        return "Running Builds";
```

(also preferably renaming class to match)

File: src/main/java/com/cloudbees/jenkins/support/impl/RunningJobs.java [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+    }
+
+    @Override
+    public void addContents(@NonNull Container result) {
+        result.add(
+            new PrefilteredPrintedContent("nodes/master/running-jobs.txt") {
```
Comment (jglick) [inline src/main/java/com/cloudbees/jenkins/support/impl/RunningJobs.java]:
```suggestion
            new PrefilteredPrintedContent("nodes/master/running-builds.txt") {
```

Comment (jglick) [inline src/main/java/com/cloudbees/jenkins/support/impl/RunningJobs.java]: Perhaps this file should be at top level? It would be pretty misleading for something under `nodes/master/` to list freestyle builds running on an agent. Those _could_ be moved to a separate `nodes/*/running-builds.txt`, though I am not sure if that is helpful or just annoying.

In the case of Pipeline builds, there is no unique associated node, though in many cases there will be exactly one `PlaceholderTask`. (There could be several, if in `parallel`, or none, if paused for example in `input`.)

File: src/test/java/com/cloudbees/jenkins/support/impl/RunningJobsTest.java [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+
+    @Test
+    public void addContents() throws Exception {
+        FreeStyleProject p = j.createFreeStyleProject(JOB_NAME);
+        QueueTaskFuture<FreeStyleBuild> build = p.scheduleBuild2(0);
+        build.waitForStart();
+
+        String output = SupportTestUtils.invokeComponentToString(new RunningJobs());
+
+        assertThat(output, containsString(p.getName()));
```
Comment (jglick) [inline src/test/java/com/cloudbees/jenkins/support/impl/RunningJobsTest.java]: Could perhaps be more specific:
```suggestion
        assertThat(output, containsString(p.getName() + "#1"));
```

or whatever it is.

(BTW not a bad idea to use `JenkinsRule.createFolder` and assert that content contains the full path.)

File: src/test/java/com/cloudbees/jenkins/support/impl/RunningJobsTest.java [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+    }
+
+    @Test
+    public void addContentsPipeline() throws Exception {
+        WorkflowJob p = j.createProject(WorkflowJob.class, JOB_NAME);
```
Comment (jglick) [inline src/test/java/com/cloudbees/jenkins/support/impl/RunningJobsTest.java]: I think you are going to need an actual script here! Or the build will just be some degenerate broken thing. Surprised this test can even pass.

```java
p.setDefinition(new CpsFlowDefinition("node {semaphore 'wait'}", true);
WorkflowRun build = p.scheduleBuild2(0).waitForStart();
SemaphoreStep.waitForStart("wait/1", build);
```

and then you want to assert that the output does contain a representation of the build but does _not_ contain a representation of the `node` block.

File: src/main/java/com/cloudbees/jenkins/support/impl/RunningJobs.java [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+                @Override
+                protected void printTo(PrintWriter out, ContentFilter filter) {
+                    Optional.ofNullable(Jenkins.getInstanceOrNull())
+                        .ifPresent(jenkins -> Arrays.stream(jenkins.getComputers())
+                            .flatMap(computer -> computer.getAllExecutors().stream())
+                            .collect(Collectors.toList())
+                            .forEach(executor -> Optional.ofNullable(executor.getCurrentExecutable())
+                                .filter(executable -> !(executable.getParent() instanceof ExecutorStepExecution.PlaceholderTask))
```
Comment (jglick) [inline src/main/java/com/cloudbees/jenkins/support/impl/RunningJobs.java]: Would I think suffice to say
```suggestion
                                .filter(executable -> executable.getParent() == executable.getParent().getOwnerTask())
```

which will filter out `PlaceholderTask`s without needing the dependency on `workflow-durable-task-step`—this could become `test`-scoped.

For reasons I do not fully understand, neither `MatrixConfiguration` nor `MavenModule` seem to override this method, so these would not be filtered out.

File: pom.xml (hunk 9 above)
Comment (Evildethow) [reply pom.xml]: yes, understood, although I'm considering it out-of-scope currently.

File: src/main/java/com/cloudbees/jenkins/support/impl/RunningJobs.java (hunk 11 above)
Comment (Evildethow) [reply src/main/java/com/cloudbees/jenkins/support/impl/RunningJobs.java]: yes, that makes sense.

File: src/main/java/com/cloudbees/jenkins/support/impl/RunningJobs.java (hunk 14 above)
Comment (Evildethow) [reply src/main/java/com/cloudbees/jenkins/support/impl/RunningJobs.java]: Interesting. I'll see what happens when I add some tests.

**Comment by Evildethow** [comment]
Thanks for the review, @jglick

**Comment by Evildethow** [comment]
Requested changes made. @aheritier I think this one is good to go.

File: src/main/java/com/cloudbees/jenkins/support/impl/RunningBuilds.java [hunk 15]
```diff
@@ -0,0 +1,52 @@
...
+        result.add(
+            new PrefilteredPrintedContent("running-builds.txt") {
+                @Override
+                protected void printTo(PrintWriter out, ContentFilter filter) {
+                    Optional.ofNullable(Jenkins.getInstanceOrNull())
+                        .ifPresent(jenkins -> Arrays.stream(jenkins.getComputers())
+                            .flatMap(computer -> computer.getAllExecutors().stream())
+                            .collect(Collectors.toList())
+                            .forEach(executor -> Optional.ofNullable(executor.getCurrentExecutable())
+                                .filter(executable -> executable.getParent() == executable.getParent().getOwnerTask())
```
Comment (varyvol) [inline src/main/java/com/cloudbees/jenkins/support/impl/RunningBuilds.java]: Is it correct filtering before asking if the executable is present?

File: src/test/java/com/cloudbees/jenkins/support/impl/RunningBuildsTest.java [hunk 16]
```diff
@@ -0,0 +1,70 @@
...
+
+    @Test
+    public void addContents() throws Exception {
+        FreeStyleProject p = j.createFreeStyleProject(JOB_NAME);
+        QueueTaskFuture<FreeStyleBuild> freeStyleBuildQueueTaskFuture = p.scheduleBuild2(0);
+        FreeStyleBuild build = freeStyleBuildQueueTaskFuture.waitForStart();
+
+        String output = SupportTestUtils.invokeComponentToString(new RunningBuilds());
+
+        assertThat(output, containsString(String.format(EXPECTED_OUTPUT_FORMAT, p.getName(), build.getNumber())));
```
Comment (varyvol) [inline src/test/java/com/cloudbees/jenkins/support/impl/RunningBuildsTest.java]: So is this eventually filtered or not?

File: src/main/java/com/cloudbees/jenkins/support/impl/RunningBuilds.java (hunk 15 above)
Comment (Evildethow) [reply src/main/java/com/cloudbees/jenkins/support/impl/RunningBuilds.java]: Yes. See https://docs.oracle.com/javase/8/docs/api/java/util/Optional.html#filter-java.util.function.Predicate-

File: src/test/java/com/cloudbees/jenkins/support/impl/RunningBuildsTest.java (hunk 16 above)
Comment (Evildethow) [reply src/test/java/com/cloudbees/jenkins/support/impl/RunningBuildsTest.java]: I don't really understand your question. Sorry.

Comment (varyvol) [reply src/test/java/com/cloudbees/jenkins/support/impl/RunningBuildsTest.java]: The test is checking the actual job name is displayed in the output string, when that should be something anonymized. Is it that the info is not anonymized or that the utility test function is not taking that into account?

Comment (Evildethow) [reply src/test/java/com/cloudbees/jenkins/support/impl/RunningBuildsTest.java]: I'll add a test for anonymization.

Comment (Evildethow) [reply src/test/java/com/cloudbees/jenkins/support/impl/RunningBuildsTest.java]: https://github.com/jenkinsci/support-core-plugin/pull/189/commits/6d986a5207a3633930088c71c9f90cacbf090018

File: src/test/java/com/cloudbees/jenkins/support/SupportTestUtils.java [hunk 17]
```diff
@@ -35,4 +38,24 @@ public void add(@CheckForNull Content content) {
 
     return baos.toString();
   }
+
+    /**
+     * Invoke a component with {@link ContentFilter}, and return the component contents as a String.
+     */
+    public static String invokeComponentToString(final Component component, final ContentFilter filter) {
+        final ByteArrayOutputStream baos = new ByteArrayOutputStream();
```
Comment (fcojfernandez) [inline src/test/java/com/cloudbees/jenkins/support/SupportTestUtils.java]:
```suggestion
        try(ByteArrayOutputStream baos = new ByteArrayOutputStream()) {
```

File: src/test/java/com/cloudbees/jenkins/support/SupportTestUtils.java [hunk 18]
```diff
@@ -35,4 +38,24 @@ public void add(@CheckForNull Content content) {
...
+                public void add(@CheckForNull Content content) {
+                    try {
+                        ((PrefilteredContent) Objects.requireNonNull(content)).writeTo(baos, filter);
+                    } catch (IOException e) {
+                        e.printStackTrace();
+                    }
+                }
+            });
+
+        return baos.toString();
```
Comment (fcojfernandez) [inline src/test/java/com/cloudbees/jenkins/support/SupportTestUtils.java]:
```suggestion
            return baos.toString();
        }
```

File: src/main/java/com/cloudbees/jenkins/support/impl/RunningBuilds.java [hunk 19]
```diff
@@ -0,0 +1,52 @@
...
+        return "Running Builds";
+    }
+
+    @Override
+    public void addContents(@NonNull Container result) {
+        result.add(
+            new PrefilteredPrintedContent("running-builds.txt") {
+                @Override
+                protected void printTo(PrintWriter out, ContentFilter filter) {
+                    Optional.ofNullable(Jenkins.getInstanceOrNull())
```
Comment (fcojfernandez) [inline src/main/java/com/cloudbees/jenkins/support/impl/RunningBuilds.java]: Why don't just use `Jenkins.get()`? It's already Nonnull.

**Comment by Evildethow** [comment]
> Plugin dependencies could be also cleaned up after moving to BOM

Understood, but I'd rather address that separately. I'm considering it out-of-scope for now.

**Comment by aheritier** [comment]
yes let's cleanup the deps in another PR @Evildethow

**Comment by Evildethow** [comment]
As there seem to be no further requests for changes I'm going to merge this.

Thanks, all.

**Comment by Evildethow** [comment]
I've created https://issues.jenkins-ci.org/browse/JENKINS-59453 to address migrating to the BOM
